keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md

```diff
@@ -0,0 +1,545 @@
+
+On startup, the tool should connect to the cluster and read the registry ConfigMap.
+
+If the ConfigMap specifies a `help` URL, the tool may prompt the user to set up
+a registry for faster development.
+
+If the ConfigMap specifies a registry location, the tool may automatically
+adjust image pushes to push to the specified registry.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+- Users may see the ConfigMap and draw mistaken conclusions about how to
+  interact with it. They think deleting the ConfigMap would delete the local
+  registry (which it does not).
+
+- This KEP does not specify any automatic mechanism for keeping the
+  RegistryStatus up-to-date with how the cluster is configured.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+Tools that configure a local registry should apply the following ConfigMap to
+the cluster:
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry
+  namespace: kube-public
+data:
+  RegistryStatus: |
+    apiVersion: registry.x-k8s.io/v1alpha1
+    kind: RegistryStatus
+    host: "localhost:5000"
+    hostFromCluster: "registry:5000"
+    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
+```
+
+Documentation that educates people on how to set up a registry manually should
+include instructions on this.
+
+The name of the config map must be `registry`.
+
+The namespace must be `kube-public`.
+
+The `RegistryStatus` field of the ConfigMap data takes the form of a component
+config, like the `kubeadm-config` ConfigMap in `kube-system`.
+
+There are 3 fields:
+
+- `host`: Advertises the host (hostname and port) of the registry, as seen from outside the cluster. 
+  This is the registry host that tools should push images to.
+  
+- `hostFromCluster`: Advertises the host (hostname and port) of the registry, as
+  seen from the cluster's container runtime. This is the registry that should
+  appear in Kubernetes objects applied to the cluster. Should be ignored if
+  `host` is not set.  If omitted, assumed to be the same as `host`.
```

Comment:
i wouldn't be surprised if core k8s is already doing this for some types, but API wise, propagating the user-fed value of `host` to `hostFromCluster` on the side of the API machinery defaulting for readers, can be a bit odd with respect to who added the value "was it the defaulting or the writer of the ConfigMap"?

so while it will remove code that readers will have to otherwise duplicate, i think it makes more sense to not do this type of defaulting and let the users manage what value to pick with - i.e. `if len(hostFromCluster) == 0 { return host }`.

Reply:
oh yes, that's what I meant to imply - the readers of this fields are expected to do this manually, rather than relying on machinery. re-worded this a bit